build-tools/packages/build-cli/src/commands/typetests.ts

```diff
@@ -153,29 +158,62 @@ export function previousVersion(version: string): string {
 
 /**
  * Updates configuration for type tests in package.json
+ *
+ * @internal
  */
-function updateTypeTestConfiguration(pkgJson: PackageJson, options: TypeTestConfigActions): void {
+export function updateTypeTestConfiguration(
+	pkgJson: PackageJson,
+	options: TypeTestConfigActions,
+): void {
 	if (options.version !== undefined) {
-		const oldDepName = `${pkgJson.name}-previous`;
+		applyTypeTestVersionOptions(pkgJson, options.version);
+	}
 
-		// Packages can explicitly opt out of type tests by setting typeValidation.disabled to true.
-		const enabled = pkgJson.typeValidation?.disabled !== true;
+	if (options.resetBroken !== undefined) {
+		resetBrokenTests(pkgJson, options.resetBroken);
+	}
+}
 
-		if (!enabled || options.version === VersionOptions.Clear) {
-			// There isn't a type safe alternative to delete here,
-			// and delete should be safe since the key can't collide with anything builtin since it has to end in `-previous` which makes it not a valid identifier.
+/**
+ * Adds or removes the devDependency on a previous version of a package thatis used for type testing. This function only
+ * affects the `devDependencies` nodes in package.json.
+ *
+ * @internal
+ */
+export function applyTypeTestVersionOptions(
+	pkgJson: PackageJson,
+	versionOptions: string | VersionOptions,
+): void {
+	const oldDepName = `${pkgJson.name}-previous`;
+
+	// Packages can explicitly opt out of type tests by setting typeValidation.disabled to true.
+	const enabled = pkgJson.typeValidation?.disabled !== true;
+
+	if (!enabled || versionOptions === VersionOptions.Clear) {
+		// There isn't a type safe alternative to delete here,
+		// and delete should be safe since the key can't collide with anything builtin since it has to end in `-previous`
+		// which makes it not a valid identifier.
+		if (pkgJson.devDependencies?.[oldDepName] !== undefined) {
 			// eslint-disable-next-line @typescript-eslint/no-dynamic-delete
 			delete pkgJson.devDependencies[oldDepName];
-		} else if (options.version !== VersionOptions.ClearIfDisabled) {
-			const newVersion: string =
-				options.version === VersionOptions.Previous
-					? previousVersion(pkgJson.version)
-					: options.version;
-			pkgJson.devDependencies[oldDepName] = `npm:${pkgJson.name}@${newVersion}`;
 		}
+	} else if (versionOptions !== VersionOptions.ClearIfDisabled) {
+		const newVersion: string =
+			versionOptions === VersionOptions.Previous
+				? previousVersion(pkgJson.version)
+				: versionOptions;
+		pkgJson.devDependencies ??= {};
+		pkgJson.devDependencies[oldDepName] = `npm:${pkgJson.name}@${newVersion}`;
 	}
+}
 
-	if (options.resetBroken === true && pkgJson.typeValidation !== undefined) {
+/**
+ * Removes any `typeValidation.broken` entries from package.json.
+ *
+ * @internal
+ */
+export function resetBrokenTests(pkgJson: PackageJson, resetBroken: boolean): void {
```

Comment:
if you want to make this easily testable, you could be more specific about what it takes in, for example After applying the above suggestion as well:
```suggestion
export function resetBrokenTests(pkgJson: { typeValidation?: ITypeValidationConfig}): void {
```

This way a test for this function could be very trivial. I think this would be enough of a test suite for such a simple function:
```
assert.deepEqual(resetBrokenTests({}, {});
assert.deepEqual(resetBrokenTests({ typeValidation: {broken: {}}}, {});
```

(I'd put both in one test for simplicity)

Reply:
This would require me to export the ITypeValidationConfig type from build-tools, which I'd really like to avoid. I guess I could copy the definition for ITypeValidationConfig, but that doesn't seem correct either.

Reply:
ITypeValidationConfig is part of the already package exported PackageJson type. Exporting a type but not the types it referenced is questionable, and API-Extractor will reject it if you use our strict configuration.

Generally forcing someone to use a type made up of parts they can't refer to because they arn't exported is bad: it makes it hard to work with the type, hard to get docs for it etc. That is why API-Extractor consideres it an error.

Also the type is effectivly already exported as `PackageJson["typeValidation"]`, and has to appear in the package d.ts file since its used in that file: you are not losing any real encapsulation by exprting it with a proper name.

Reply:
Good points.

Reply:
I exported the typetest related types, but there are others that should be exported. I'll leave those for a future change.